File: go/trivia.go
package main

import (
	"fmt"
	"math/rand"
	"time"
)

type Game struct {
	players      []string
	places       []int
	purses       []int
	inPenaltyBox []bool

	popQuestions     []string
	scienceQuestions []string
	sportsQuestions  []string
	rockQuestions    []string

	currentPlayer            int
	isGettingOutOfPenaltyBox bool
}

func NewGame() *Game {
	game := &Game{}
	for i := 0; i < 6; i++ {
		game.places = append(game.places, 0)
		game.purses = append(game.purses, 0)
		game.inPenaltyBox = append(game.inPenaltyBox, false)
	}

	for i := 0; i < 50; i++ {
		game.popQuestions = append(game.popQuestions,
			fmt.Sprintf("Pop Question %d\n", i))
		game.scienceQuestions = append(game.scienceQuestions,
			fmt.Sprintf("Science Question %d\n", i))
		game.sportsQuestions = append(game.sportsQuestions,
			fmt.Sprintf("Sports Question %d\n", i))
		game.rockQuestions = append(game.rockQuestions,
			game.CreateRockQuestion(i))
	}

	return game
}

func (me *Game) CreateRockQuestion(index int) string {
	return fmt.Sprintf("Rock Question %d\n", index)
}

func (me *Game) IsPlayable() bool {
	return me.howManyPlayers() >= 2
}

func (me *Game) howManyPlayers() int {
	return len(me.players)
}

func (me *Game) Add(playerName string) bool {
	me.players = append(me.players, playerName)
	me.places[me.howManyPlayers()] = 0
	me.purses[me.howManyPlayers()] = 0
	me.inPenaltyBox[me.howManyPlayers()] = false

	fmt.Printf("%s was added\n", playerName)
	fmt.Printf("They are player number %d\n", len(me.players))

	return true
}

func (me *Game) Roll(roll int) {
	fmt.Printf("%s is the current player\n", me.players[me.currentPlayer])
	fmt.Printf("They have rolled a %d\n", roll)

	if me.inPenaltyBox[me.currentPlayer] {
		if roll%2 != 0 {
			me.isGettingOutOfPenaltyBox = true

			fmt.Printf("%s is getting out of the penalty box\n", me.players[me.currentPlayer])
			me.places[me.currentPlayer] = me.places[me.currentPlayer] + roll
			if me.places[me.currentPlayer] > 11 {
				me.places[me.currentPlayer] = me.places[me.currentPlayer] - 12
			}

			fmt.Printf("%s's new location is %d\n", me.players[me.currentPlayer], me.places[me.currentPlayer])
			fmt.Printf("The category is %s\n", me.currentCategory())
			me.askQuestion()
		} else {
			fmt.Printf("%s is not getting out of the penalty box\n", me.players[me.currentPlayer])
			me.isGettingOutOfPenaltyBox = false
		}
	} else {
		me.places[me.currentPlayer] = me.places[me.currentPlayer] + roll
		if me.places[me.currentPlayer] > 11 {
			me.places[me.currentPlayer] = me.places[me.currentPlayer] - 12
		}

		fmt.Printf("%s's new location is %d\n", me.players[me.currentPlayer], me.places[me.currentPlayer])
		fmt.Printf("The category is %s\n", me.currentCategory())
		me.askQuestion()
	}
}

func (me *Game) askQuestion() {
	if me.currentCategory() == "Pop" {
		question := me.popQuestions[0]
		me.popQuestions = me.popQuestions[1:]
		fmt.Printf(question)
	}
	if me.currentCategory() == "Science" {
		question := me.scienceQuestions[0]
		me.scienceQuestions = me.scienceQuestions[1:]
		fmt.Printf(question)
	}
	if me.currentCategory() == "Sports" {
		question := me.sportsQuestions[0]
		me.sportsQuestions = me.sportsQuestions[1:]
		fmt.Printf(question)
	}
	if me.currentCategory() == "Rock" {
		question := me.rockQuestions[0]
		me.rockQuestions = me.rockQuestions[1:]
		fmt.Printf(question)
	}
}

func (me *Game) currentCategory() string {
	if me.places[me.currentPlayer] == 0 {
		return "Pop"
	}
	if me.places[me.currentPlayer] == 4 {
		return "Pop"
	}
	if me.places[me.currentPlayer] == 8 {
		return "Pop"
	}
	if me.places[me.currentPlayer] == 1 {
		return "Science"
	}
	if me.places[me.currentPlayer] == 5 {
		return "Science"
	}
	if me.places[me.currentPlayer] == 9 {
		return "Science"
	}
	if me.places[me.currentPlayer] == 2 {
		return "Sports"
	}
	if me.places[me.currentPlayer] == 6 {
		return "Sports"
	}
	if me.places[me.currentPlayer] == 10 {
		return "Sports"
	}
	return "Rock"
}

func (me *Game) WasCorrectlyAnswered() bool {
	if me.inPenaltyBox[me.currentPlayer] {
		if me.isGettingOutOfPenaltyBox {
			fmt.Println("Answer was correct!!!!")
			me.purses[me.currentPlayer] += 1
			fmt.Printf("%s now has %d Gold Coins.\n", me.players[me.currentPlayer], me.purses[me.currentPlayer])

			winner := me.didPlayerWin()
			me.currentPlayer += 1
			if me.currentPlayer == len(me.players) {
				me.currentPlayer = 0
			}

			return winner
		} else {
			me.currentPlayer += 1
			if me.currentPlayer == len(me.players) {
				me.currentPlayer = 0
			}
			return true
		}
	} else {

		fmt.Println("Answer was corrent!!!!")
		me.purses[me.currentPlayer] += 1
		fmt.Printf("%s now has %d Gold Coins.\n", me.players[me.currentPlayer], me.purses[me.currentPlayer])

		winner := me.didPlayerWin()
		me.currentPlayer += 1
		if me.currentPlayer == len(me.players) {
			me.currentPlayer = 0
		}

		return winner
	}

	return false
}

func (me *Game) didPlayerWin() bool {
	return !(me.purses[me.currentPlayer] == 6)
}

func (me *Game) WrongAnswer() bool {
	fmt.Println("Question was incorrectly answered")
	fmt.Printf("%s was sent to the penalty box\n", me.players[me.currentPlayer])
	me.inPenaltyBox[me.currentPlayer] = true

	me.currentPlayer += 1
	if me.currentPlayer == len(me.players) {
		me.currentPlayer = 0
	}

	return true
}

func main() {
	notAWinner := false

	game := NewGame()

	game.Add("Chet")
	game.Add("Pat")
	game.Add("Sue")

	rand.Seed(time.Now().UTC().UnixNano())

	for {
		game.Roll(rand.Intn(5) + 1)

		if rand.Intn(9) == 7 {
			notAWinner = game.WrongAnswer()
		} else {
			notAWinner = game.WasCorrectlyAnswered()

		}

		if !notAWinner {
			break
		}
	}
}

File: perl/trivia.pl
#!/usr/bin/perl

package Game;

sub new {
    my ($class) = @_;
    my $self = bless( {}, $class );

    $self->{players}        = [];
    $self->{places}         = [ (0) x 6 ];
    $self->{purses}         = [ (0) x 6 ];
    $self->{in_penalty_box} = [ (0) x 6 ];

    $self->{pop_questions}     = [];
    $self->{science_questions} = [];
    $self->{sports_questions}  = [];
    $self->{rock_questions}    = [];

    $self->{current_player}                = 0;
    $self->{is_getting_out_of_penalty_box} = 0;

    for ( my $i = 0; $i < 50; $i++ ) {
        push( @{ $self->{pop_questions} },     "Pop Question $i" );
        push( @{ $self->{science_questions} }, "Science Question $i" );
        push( @{ $self->{sports_questions} },  "Sports Question $i" );
        push( @{ $self->{rock_questions} }, $self->create_rock_question($i) );
    }

    return $self;
}

sub create_rock_question {
    my ( $self, $index ) = @_;
    return "Rock Question $index";
}

sub is_playable {
    my ($self) = @_;
    return $self->how_many_players() >= 2;
}

sub add {
    my ( $self, $player_name ) = @_;
    push( @{ $self->{players} }, $player_name );
    ${ $self->{places} }[ $self->how_many_players() ]         = 0;
    ${ $self->{purses} }[ $self->how_many_players() ]         = 0;
    ${ $self->{in_penalty_box} }[ $self->how_many_players() ] = 0;

    print $player_name . " was added\n";
    print "They are player number " . scalar( @{ $self->{players} } ) . "\n";

    return 1;
}

sub how_many_players {
    my ($self) = @_;
    return scalar( @{ $self->{players} } );
}

sub roll {
    my ( $self, $roll ) = @_;
    print ${ $self->{players} }[ $self->{current_player} ]
        . " is the current player\n";
    print "They have rolled a " . $roll . "\n";

    if ( ${ $self->{in_penalty_box} }[ $self->{current_player} ] ) {
        if ( $roll % 2 != 0 ) {
            $self->{is_getting_out_of_penalty_box} = 1;

            print ${ $self->{players} }[ $self->{current_player} ]
                . " is getting out of the penalty box\n";
            ${ $self->{places} }[ $self->{current_player} ]
                = ${ $self->{places} }[ $self->{current_player} ] + $roll;
            if ( ${ $self->{places} }[ $self->{current_player} ] > 11 ) {
                ${ $self->{places} }[ $self->{current_player} ]
                    = ${ $self->{places} }[ $self->{current_player} ] - 12;
            }
            print ${ $self->{players} }[ $self->{current_player} ]
                . "'s new location is "
                . ${ $self->{places} }[ $self->{current_player} ] . "\n";
            print "The category is " . $self->_current_category() . "\n";
            $self->_ask_question();
        }
        else {
            print ${ $self->{players} }[ $self->{current_player} ]
                . " is not getting out of the penalty box\n";
            $self->{is_getting_out_of_penalty_box} = 0;
        }
    }
    else {
        ${ $self->{places} }[ $self->{current_player} ]
            = ${ $self->{places} }[ $self->{current_player} ] + $roll;
        if ( ${ $self->{places} }[ $self->{current_player} ] > 11 ) {
            ${ $self->{places} }[ $self->{current_player} ]
                = ${ $self->{places} }[ $self->{current_player} ] - 12;
        }

        print ${ $self->{players} }[ $self->{current_player} ]
            . "'s new location is "
            . ${ $self->{places} }[ $self->{current_player} ] . "\n";
        print "The category is " . $self->_current_category() . "\n";
        $self->_ask_question();
    }
}

sub _ask_question {
    my ($self) = @_;
    if ( $self->_current_category() eq 'Pop' ) {
        print pop( @{ $self->{pop_questions} } ) . "\n";
    }
    if ( $self->_current_category() eq 'Science' ) {
        print pop( @{ $self->{science_questions} } ) . "\n";
    }
    if ( $self->_current_category() eq 'Sports' ) {
        print pop( @{ $self->{sports_questions} } ) . "\n";
    }
    if ( $self->_current_category() eq 'Rock' ) {
        print pop( @{ $self->{rock_questions} } ) . "\n";
    }
}

sub _current_category {
    my ($self) = @_;
    return 'Pop' if ( ${ $self->{places} }[ $self->{current_player} ] == 0 );
    return 'Pop' if ( ${ $self->{places} }[ $self->{current_player} ] == 4 );
    return 'Pop' if ( ${ $self->{places} }[ $self->{current_player} ] == 8 );
    return 'Science'
        if ( ${ $self->{places} }[ $self->{current_player} ] == 1 );
    return 'Science'
        if ( ${ $self->{places} }[ $self->{current_player} ] == 5 );
    return 'Science'
        if ( ${ $self->{places} }[ $self->{current_player} ] == 9 );
    return 'Sports'
        if ( ${ $self->{places} }[ $self->{current_player} ] == 2 );
    return 'Sports'
        if ( ${ $self->{places} }[ $self->{current_player} ] == 6 );
    return 'Sports'
        if ( ${ $self->{places} }[ $self->{current_player} ] == 10 );
    return 'Rock';
}

sub was_correctly_answered {
    my ($self) = @_;
    if ( ${ $self->{in_penalty_box} }[ $self->{current_player} ] ) {
        if ( $self->{is_getting_out_of_penalty_box} ) {
            print "Answer was correct!!!!\n";
            ${ $self->{purses} }[ $self->{current_player} ] += 1;
            print ${ $self->{players} }[ $self->{current_player} ]
                . " now has "
                . ${ $self->{purses} }[ $self->{current_player} ]
                . " Gold Coins.\n";

            my $winner = $self->_did_player_win();
            $self->{current_player} += 1;
            if ( $self->{current_player} == scalar( @{ $self->{players} } ) )
            {
                $self->{current_player} = 0;
            }

            return $winner;
        }
        else {
            $self->{current_player} += 1;
            if ( $self->{current_player} == scalar( @{ $self->{players} } ) )
            {
                $self->{current_player} = 0;
            }
            return 1;
        }
    }
    else {
        print "Answer was corrent!!!!\n";
        ${ $self->{purses} }[ $self->{current_player} ] += 1;
        print ${ $self->{players} }[ $self->{current_player} ]
            . " now has "
            . ${ $self->{purses} }[ $self->{current_player} ]
            . " Gold Coins.\n";

        my $winner = $self->_did_player_win();
        $self->{current_player} += 1;
        if ( $self->{current_player} == scalar( @{ $self->{players} } ) ) {
            $self->{current_player} = 0;
        }

        return $winner;
    }
}

sub wrong_answer {
    my ($self) = @_;
    print "Question was incorrectly answered\n";
    print ${ $self->{players} }[ $self->{current_player} ]
        . " was sent to the penalty box\n";
    ${ $self->{in_penalty_box} }[ $self->{current_player} ] = 1;

    $self->{current_player} += 1;
    if ( $self->{current_player} == scalar( @{ $self->{players} } ) ) {
        $self->{current_player} = 0;
    }
    return 1;
}

sub _did_player_win {
    my ($self) = @_;
    return not( ${ $self->{purses} }[ $self->{current_player} ] == 6 );
}

package GameRunner;

my $not_a_winner = 0;

my $game = Game->new();

$game->add('Chet');
$game->add('Pat');
$game->add('Sue');

while (1) {
    $game->roll( int( rand(5) ) + 1 );

    if ( int( rand(9) ) == 7 ) {
        $not_a_winner = $game->wrong_answer();
    }
    else {
        $not_a_winner = $game->was_correctly_answered();
    }

    if ( not $not_a_winner ) {
        last;
    }
}
